### ISSUE TITLE
Enable routing by integrating with Graphhopper for San Ramon

### ISSUE THREAD

**Issue by neelamsaboo** [usdigitalresponse/neighbor-express#149]

**Comment by geoffreylitt** [comment]
# Current status

We have a solution working which clusters addresses into groups.

Based on our last feedback call w/ San Ramon, the pending changes are the following:

* put address in _order_ within each cluster for an efficient driving route
* generate a map view that can be printed to show the route (eg if the volunteer doesn't have a smartphone)

# Progress

I investigated a few options for solving the problems above. The main idea is to hopefully link to an external mapping service that can do routing/mapping so we do as little work as possible beyond grouping addresses.

## Using MapQuest

San Ramon currently manually fills in MapQuest Route Planner for routing, so it would be awesome if we could essentially auto-generate the input they need for that.

MapQuest has [multiple](https://developer.mapquest.com/documentation/tools/link-to-mapquest/) [dev tools](https://developer.mapquest.com/documentation/tools/link-to-route-planner/) which purport to allow generating links to route planner and passing in addresses. This would be exactly what we want -- we could generate URLs in airtable for each route which link out to MapQuest; MapQuest would order the route and generate a map. Unfortunately all of these tools seem to not work in practice. MapQuest seems generally poorly maintained and documented.

A potentially workable option for sticking with Mapquest would be to autogenerate CSV output from Airtable that can be manually copy-pasted into the Mapquest tool. Mapquest can reorder addresses on their end so we could basically just reformat our current existing clusters into their import format.

<img width="389" alt="Screen Shot 2020-08-09 at 11 53 02 AM" src="https://user-images.githubusercontent.com/934016/89736370-eb5c3800-da36-11ea-9740-21db6ea25791.png">

This might require a bit of manual effort (one copy-paste from Airtable to Mapquest per route, essentially), but it would probably still save time over their current process, and would keep them in the tool they're familiar with.

## Google Maps

Google Maps has a [much better setup](https://developers.google.com/maps/documentation/urls/get-started#directions-examples-using-waypoints) for linking to a directions view, which could allow for a more automated flow. There could be one link from Airtable for each route which would go right to a Google Maps directions view.

The problem is, GMaps requires the addresses to already be in the right order. We don't have an ordering yet. We would need to adjust our usage of the Graphhopper API to generate address ordering. Not sure how hard this is.

# Next steps

Consider these two paths forward, get feedback. I'm inclined to start w/ fleshing out the mapquest copy-paste option--I suspect it will be less work for us and less change for them.

If we go that route, next step is to try prototyping a CSV import/export workflow, record a quick demo, and send to them for feedback. I think that could probably be done this week.

**Comment by geoffreylitt** [comment]
thanks so much for offering to look into this @alienngator @michaelogren !

a few other pointers to context:

* the existing code for the routing stuff is in a scripting block -- the version in San Ramon's live Airtable base is most up to date. There's also a scripting block in staging which should have roughly the same content.
* We currently use Graphhopper's Cluster API because it seemed most convenient for the task initially. But there are [other APIs](https://docs.graphhopper.com/) available which might be useful.

One issue I realized with the copy-paste-to-Mapquest option here is that Mapquest needs the different components of the address separated out in the CSV import. 

We don't currently have this in Airtable, but it wouldn't be that hard to populate. We already call the Graphhopper Geocoding API in the routing block to turn the raw address string into more structured data. Currently we just store the latitude/longitude as shown below. But Graphhopper also returns us other data that we could store, like street address, zipcode, etc.

```js
let address = record.getCellValue("Full Address");
    let geocoded = await fetch(
      `https://graphhopper.com/api/1/geocode?key=${GRAPHHOPPER_KEY}&q=${address}&provider=nominatim&point=${SANRAMON_LATLONG}`
    )
    let result = await geocoded.json(); 
    if (result.hits.length > 0) {
      let latlong = result.hits[0].point;
      let latlongSerialized = JSON.stringify(latlong);
      await deliveriesTable.updateRecordAsync(record, {
        "Latlong (internal)": latlongSerialized
      })
```

**Comment by alienngator** [comment]
TL;DR: Did some investigation into all options Geoff mentioned above. I also think that the manual copy-paste option has the best value-to-effort ratio. Notes below.

### Option 1: Google Maps
**1. 1. Using Maps URLs** (i.e. we generate links that can be opened directly in Google Maps, given the stops on the route).
What users see: One such link for each route.
Pros: Doesn’t require integration with a new API.
Cons: The addresses have to be sorted first. I tried using a different Grasshopper API for this, which kinda gets the job done but isn’t the most intuitive (hence potentially harder to implement & maintain down the road) ⟶ more work for us.

**1.2. Using the Maps API** (i.e. integrate programmatically with the Google Maps Javascript API, specifically DirectionsService & DirectionsRenderer).
What users see: Routes visualized on Google Maps. I think.
Pros:
- Sorting routes on the route is trivial. There’s an “optimizeWaypoints” bit that we can flip, so that a route can be planned with stops automatically reordered.

Cons:
- By far the most work for us. Need to add the API as a direct dependency (not sure how hard this would be from an Airtable script, or if it’s even possible?).
- How to visualize the results we get from DirectionsService: There’s probably a simple way to do this if the Maps API integration is already set up, but I’m not sure.
- Paid service, requires billing.

### Option 2: Explore a different Grasshopper API
I tried using the Route Optimization API.
What users see: Routes visualized on some kind of map.
Pros:
- This API divides the stop into ordered routes, in a single API call (i.e. cluster AND sort. Currently we use the Cluster API first to cluster, *then look for other ways to sort and visualize the results).

Cons:
- We still need another map service to view the routes.
- The API itself isn’t the most intuitive & documentation is a little sparse. Future maintainability might be an issue.
- Might just be me, but the API endpoint seems a little slow from the few times I’ve tried.

### Option 3: MapQuest copy-paste
We don't even need to go through a csv file (i.e. do away with the import/export csv flow). We can just print each route directly into the Airtable output. Users can then copy that as is into MapQuest -- I'll try to upload a demo asap. (@geoffreylitt please let me know if there's a reason we shouldn't do this).
What users see: Routes visualized on MapQuest.
Pros:
- As Geoff says, “would keep users in the tool they're familiar with.”
- Least work for us, only involves a few more lines of code.

Cons:
- As Geoff says, “might require a bit of manual effort (one copy-paste from Airtable to Mapquest per route, essentially), but it would probably still save time over their current process.”

Next steps for me:
- Upload a demo for the MapQuest copy-paste option to get everyone's feedback.
- While prototyping, try to see if we can directly generate MapQuest links (hence eliminate the manual copy-paste step for users). I haven't had any success with this so far, but I haven't been trying super hard either.

**Comment by alienngator** [comment]
Side note (This is independent of any routing solution we choose, so long as Grasshopper is still used for geocoding+clustering. We should probably file a separate issue for this, but taking notes here for now):
I realized that there are accuracy issues with the geocoding results returned by Grasshopper:
- A small percentage are wildly off (e.g. all these addresses are supposed to be in San Ramon, but there are results in the East Coast or in Tokyo). This definitely affects routing results.
- A lot, if not most, are a teeny bit off (correct street, but house numbers are off by 2-3). This wouldn't affect routing, but we should keep in mind (e.g. show volunteers the user-provided addresses, instead of results returned by Grasshopper).

One getaround is to manually go through all the addresses and check the stored lat/long values. This is painful, but only a one-off task.
